plugins/metric_links/src/Metric.php

```diff
@@ -163,32 +149,32 @@ public function scan($uri, \DOMXPath $xpath, $depth, Page $page, Metrics $contex
      * This method will find broken links and mark a page appropriately
      * 
      * @param Page $page the page to mark
-     * @param array $links an array of the links found on the page
      */
-    public function markPage(Page $page, $links)
+    public function markPage(Page $page)
     {
-        $statuses = $this->getStatuses(array_unique($links));
-        $counts = array_count_values($links);
-
-        foreach ($statuses as $url=>$status) {
-            if (!$this->isError($status)) {
+        $links = $page->getLinks();
+        
+        foreach ($links as $link) {
+            if (!$this->isError($link)) {
                 //don't mark it...
                 continue;
             }
 
-            $machine_name = $this->getMachineNameForStatus($status);
+            $machine_name = $this->getMachineNameForStatus($link);
```

Comment:
These lines (163-168) seem like a good candidate for refactoring into a separate method that generates a `Mark` given a `Link`. All of the other methods in this block depend on the contents of the link object. This also makes me question why these methods have a `public` scope. Are these API endpoints needed in foreign objects?